src/manager.rs

```diff
@@ -0,0 +1,532 @@
+    SUI(SetUpdateInfo),
+    AUI(AddUpdateInfo),
+}
+
+pub trait ReadHolder<V: SelectValue> {
+    fn get_value(&self) -> Result<Option<&V>, RedisError>;
+}
+
+pub trait WriteHolder<O: Clone, V: SelectValue> {
+    fn delete(&mut self) -> Result<(), RedisError>;
+    fn get_value(&self) -> Result<Option<&mut V>, RedisError>;
+    fn set_value(&mut self, path: Vec<String>, v: O) -> Result<bool, RedisError>;
+    fn dict_add(&mut self, path: Vec<String>, key: &str, v: O) -> Result<bool, RedisError>;
+    fn delete_path(&mut self, path: Vec<String>) -> Result<bool, RedisError>;
+    fn incr_by(&mut self, path: Vec<String>, num: &str) -> Result<Number, RedisError>;
+    fn mult_by(&mut self, path: Vec<String>, num: &str) -> Result<Number, RedisError>;
+    fn pow_by(&mut self, path: Vec<String>, num: &str) -> Result<Number, RedisError>;
+    fn bool_toggle(&mut self, path: Vec<String>) -> Result<bool, RedisError>;
+    fn str_append(&mut self, path: Vec<String>, val: String) -> Result<usize, RedisError>;
+    fn arr_append(&mut self, path: Vec<String>, args: Vec<O>) -> Result<usize, RedisError>;
+    fn arr_insert(
+        &mut self,
+        path: Vec<String>,
+        args: &Vec<O>,
+        index: i64,
+    ) -> Result<usize, RedisError>;
+    fn arr_pop(&mut self, path: Vec<String>, index: i64) -> Result<Option<String>, RedisError>;
+    fn arr_trim(&mut self, path: Vec<String>, start: i64, stop: i64) -> Result<usize, RedisError>;
+    fn clear(&mut self, path: Vec<String>) -> Result<usize, RedisError>;
+    fn apply_changes(&mut self, ctx: &Context, command: &str) -> Result<(), RedisError>;
+}
+
+pub trait Manager {
+    /* V - SelectValue that the json path library can work on
+     * O - SelectValue Holder
+     * Naiv implementation is that V and O are from the same type but its not
+     * always possible so they are seperated
+     */
+    type V: SelectValue;
+    type O: Clone;
+    type WriteHolder: WriteHolder<Self::O, Self::V>;
+    type ReadHolder: ReadHolder<Self::V>;
+    fn open_key_read(&self, ctx: &Context, key: &str) -> Result<Self::ReadHolder, RedisError>;
+    fn open_key_write(&self, ctx: &Context, key: &str) -> Result<Self::WriteHolder, RedisError>;
+    fn from_str(&self, val: &str, format: Format) -> Result<Self::O, Error>;
+    fn get_memory(&self, v: &Self::V) -> Result<usize, RedisError>;
+}
+
+fn err_json(value: &Value, expected_value: &'static str) -> Error {
+    Error::from(format!(
+        "ERR wrong type of path value - expected {} but found {}",
+        expected_value,
+        RedisJSON::value_name(value)
+    ))
+}
+
+pub struct KeyHolderWrite {
+    key: RedisKeyWritable,
+    key_name: String,
+}
+
+impl KeyHolderWrite {
+    fn update<F: FnMut(Value) -> Result<Option<Value>, Error>>(
+        &self,
+        path: &Vec<String>,
+        root: &mut Value,
+        mut func: F,
+    ) -> Result<(), Error> {
+        let mut target = root;
+
+        let last_index = path.len().saturating_sub(1);
```

Comment:
Saw `saturating_sub` is nightly-only experimental API.
Might fail for users which do not have nightly rust?

Reply:
I do not use a nightly build, I guess it uses this: https://doc.rust-lang.org/std/primitive.u32.html#method.saturating_sub